### DISCUSSION TITLE
Nats self publish a scheduled cron message?

### DISCUSSION THREAD

**Discussion by NOT-HAL9000** [nats-io/nats-server#3384] [General]
I'm looking to publish a message to a stream at a certain interval.

I can set up a cron job and publish a message, however I'm wondering if it's possible (or planned) to be able to schedule a message to have the nats server self publish messages at set intervals? Basically a light implementation of cron jobs on top of jetstream. 

Thanks

**Comment by derekcollison** [comment]
At this time you will need to figure out a way to publish yourself.

We have had folks use TTL of a stream with max deliver of 1 and advisory for when the message can no longer be delivered, but this seems contrived at the moment.

**Reply by drewcorlin1** [reply]
Would it work to do the following?
* Have an ackWait that is the difference between time of publishing and the desired delivery time
* On delivery 1 (using redelivery count) ignore the message and don't ack it. 
* On delivery 2 actually do whatever you want to with the message.

**Reply by derekcollison** [reply]
Yes, that could work.

**Reply by ashupednekar** [reply]
Here's a POC based on this approach

https://github.com/ashupednekar/natshed

**Comment by Zetanova** [comment]
Beside KV, a managed job scheduler would be a great addition to nats.
It is nearly always required to also track job success / failure beside the response messages.

**Reply by jnmoyne** [reply]
You may want to take a look at https://github.com/choria-io/asyncjobs

**Reply by Zetanova** [reply]
yes, something like it, but without following:
- requires a separate process for the supervisor
- It uses the simple request/reply pattern. It breaks the NATS request/reply-many to _INBOX pattern.  

Of course can the scheduling be done with an external protocol/process, 
but it is a very basic feature for post applications like the KV store. 

I only search for something basic like:
- `schedule job A with message P in 10min to subject X.Y.Z`  NATS will then send this message in >10min to the subject as a Request/Reply  but with a special **durable _INBOX** (Inbox on JS) and tagged with Job headers
- `subscribe on job A`this will dequeue all response message (one or more) from the **durable _INBOX** tagged with Job headers
- `get jobs` regular request to get the status of current jobs

**Reply by ashupednekar** [reply]
I'd like to add this to rust, python, go... maybe through orbit or sth. Thoughts?

**Reply by neilalexander** [reply]
Think Orbit might be the right place for something like this, WDYT @Jarema?

**Reply by Zetanova** [reply]
Is this the place [orbit repo](https://github.com/synadia-io/orbit.go) ?

**Comment by tamis-laan** [comment]
Something like this would be sooooo helpful:

``` python
ns = await nats.connect("localhost:4222")

ns.cron.schedule("my_periodic_task_id", "*/1 * * * *", "my_topic",  "my message".encode())

ns.cron.schedule("my_one-shot_task_id", "2025-04-20T18:40:00+02:00", "my_topic",  "my message".encode())

ns.cron.remove("my_periodic_task_id")
```

When these tasks are one shot they should be automatically be removed after triggering and sending their message to the target topic.

**Reply by ashupednekar** [reply]
I have something similar at work, working pretty well. Any guidelines on what's accepted in orbit? Would this be okay... and what's the merge process like?

So you want a function to be invoked at the appropriate times, correct? Or is it just a scheduled produce call?

What I have is rust based, but discovers and invokes python functions, This would be different, so won't be a problem with my employer's OSS stance. 

not sure how I'd go about the discovery part for static languages, let's see. If it's only produce, I can roll this out for go/rust in a week

Lmk if there's interest. Happy to hop on it

**Comment by ripienaar** [comment]
If we added a feature where if you put a message into a stream and it has - in a header - a cron like schedule that will cause republishes into the stream at that interval. 

Would that be enough of a LEGO block for you to do what you want?

messages go into the stream so you could push consume for parallel concurrent schedule fires, pull consume for one-in-a-group doing the work etc. 

Basically just the schedule timer as a service that would combine with our other features.

**Reply by ashupednekar** [reply]
Here's how I went about it, in a nutshell
- have an internal consumer that dynamically spins up and manages task consumers
- create consumer for each task with calculated ackwait 
- mutate this ackwait based on subsequent cron or other calculations 
- run the task function, or produce in this case
- delete the consumer upon task completion, when the end clause is reached

**Reply by tamis-laan** [reply]
@ripienaar Well cron schedules are for repeating tasks. But we might also want to also schedule a single message later. So maybe a header called `Schedule` with either a cron schedule or a specific date and time?

**Reply by ripienaar** [reply]
Yes I didn’t mention it but the draft design does have a single invocation only option.

**Reply by CVMarkBradley** [reply]
Is the mentioned draft design available to review anywhere and/or is there any progress/timeline on this that can be provided?

**Reply by MauriceVanVeen** [reply]
The design can be found here: https://github.com/nats-io/nats-architecture-and-design/pull/350
The implementation for delayed messages is here: https://github.com/nats-io/nats-server/pull/7170
It will be released as part of a release preview pretty soon, otherwise our nightly builds should already contain all of this.

**Reply by CVMarkBradley** [reply]
Thanks.
Delayed messages is only one part of the design for this. Is there any information or timeline for the recurring/cron aspect of the design?

**Reply by ripienaar** [reply]
Maybe in the next release cycle - we release in 6 monthly chunks - but will see how prioritisation works out for that.
